File: java/fory-core/src/main/java/org/apache/fory/type/Float16.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.type;

import java.io.Serializable;

/**
 * IEEE 754 binary16 半精度浮点数（16位）.
 *
 * <p>格式：1位符号 + 5位指数 + 10位尾数
 *
 * <ul>
 *   <li>指数偏移：15
 *   <li>范围：±65504（最大正规数）
 *   <li>最小正规数：2^-14 ≈ 6.104e-5
 *   <li>最小次正规数：2^-24 ≈ 5.96e-8
 * </ul>
 *
 * <p>此类是不可变的，线程安全。所有算术运算通过提升到 float32 执行，然后舍入回 float16。
 *
 * <p>转换使用 IEEE 754 round-to-nearest-even 舍入模式。
 */
public final class Float16 extends Number implements Comparable<Float16>, Serializable {
  private static final long serialVersionUID = 1L;

  // ========== 位掩码常量 ==========
  private static final int SIGN_MASK = 0x8000;
  private static final int EXP_MASK = 0x7C00;
  private static final int MANT_MASK = 0x03FF;

  // ========== 特殊值位模式 ==========
  private static final short BITS_NAN = (short) 0x7E00; // 标准静默 NaN
  private static final short BITS_POS_INF = (short) 0x7C00; // +Inf
  private static final short BITS_NEG_INF = (short) 0xFC00; // -Inf
  private static final short BITS_NEG_ZERO = (short) 0x8000; // -0
  private static final short BITS_MAX = (short) 0x7BFF; // 65504
  private static final short BITS_ONE = (short) 0x3C00; // 1.0
  private static final short BITS_MIN_NORMAL = (short) 0x0400; // 2^-14
  private static final short BITS_MIN_VALUE = (short) 0x0001; // 最小次正规数

  // ========== 公共常量 ==========
  /** 非数字（Not-a-Number）值. */
  public static final Float16 NaN = new Float16(BITS_NAN);

  /** 正无穷大. */
  public static final Float16 POSITIVE_INFINITY = new Float16(BITS_POS_INF);

  /** 负无穷大. */
  public static final Float16 NEGATIVE_INFINITY = new Float16(BITS_NEG_INF);

  /** 正零. */
  public static final Float16 ZERO = new Float16((short) 0);

  /** 负零. */
  public static final Float16 NEGATIVE_ZERO = new Float16(BITS_NEG_ZERO);

  /** 值为 1.0 的常量. */
  public static final Float16 ONE = new Float16(BITS_ONE);

  /** 最大有限值 65504. */
  public static final Float16 MAX_VALUE = new Float16(BITS_MAX);

  /** 最小正规数 2^-14. */
  public static final Float16 MIN_NORMAL = new Float16(BITS_MIN_NORMAL);

  /** 最小正值（次正规数）2^-24. */
  public static final Float16 MIN_VALUE = new Float16(BITS_MIN_VALUE);

  /** Float16 的位数. */
  public static final int SIZE_BITS = 16;

  /** Float16 的字节数. */
  public static final int SIZE_BYTES = 2;

  // ========== 存储字段 ==========
  private final short bits;

  // ========== 构造方法 ==========

  private Float16(short bits) {
    this.bits = bits;
  }

  /**
   * 从位模式创建 Float16.
   *
   * @param bits IEEE 754 binary16 位模式
   * @return Float16 实例
   */
  public static Float16 fromBits(short bits) {
    return new Float16(bits);
  }

  /**
   * 从 float 值创建 Float16（带舍入）.
   *
   * <p>使用 IEEE 754 round-to-nearest-even 舍入模式。超出范围的值会舍入到 ±Infinity。
   *
   * @param value float 值
   * @return Float16 实例
   */
  public static Float16 valueOf(float value) {
    return new Float16(floatToFloat16Bits(value));
  }

  /**
   * 获取位模式.
   *
   * @return IEEE 754 binary16 位模式
   */
  public short toBits() {
    return bits;
  }

  // ========== IEEE 754 转换算法 ==========

  /**
   * 将 float32 转换为 float16 位模式。使用 round-to-nearest-even 舍入模式。
   *
   * <p>参考 Go 实现：go/fory/float16/float16.go
   */
  private static short floatToFloat16Bits(float f32) {
    int bits32 = Float.floatToRawIntBits(f32);
    int sign = (bits32 >>> 31) & 0x1;
    int exp = (bits32 >>> 23) & 0xFF;
    int mant = bits32 & 0x7FFFFF;

    int outSign = sign << 15;
    int outExp;
    int outMant;

    // 特殊值：NaN 或 Inf
    if (exp == 0xFF) {
      outExp = 0x1F;
      if (mant != 0) {
        // NaN：保留尾数高位，确保至少有一位
        outMant = 0x200 | ((mant >>> 13) & 0x1FF);
        if (outMant == 0x200) {
          outMant = 0x201; // 确保至少一位非零
        }
      } else {
        // Inf
        outMant = 0;
      }
    }
    // 零或次正规 float32
    else if (exp == 0) {
      outExp = 0;
      outMant = 0;
    }
    // 正规数
    else {
      int newExp = exp - 127 + 15;

      // 上溢到 Inf
      if (newExp >= 31) {
        outExp = 0x1F;
        outMant = 0;
      }
      // 下溢到次正规或零
      else if (newExp <= 0) {
        // 隐式 1 + 尾数
        int fullMant = mant | 0x800000;
        int shift = 1 - newExp;
        int netShift = 13 + shift;

        if (netShift >= 24) {
          // 太小，变为零
          outExp = 0;
          outMant = 0;
        } else {
          outExp = 0;
          // 舍入到最近偶数
          int roundBit = (fullMant >>> (netShift - 1)) & 1;
          int sticky = fullMant & ((1 << (netShift - 1)) - 1);
          outMant = fullMant >>> netShift;

          if (roundBit == 1 && (sticky != 0 || (outMant & 1) == 1)) {
            outMant++;
          }
        }
      }
      // 正规范围
      else {
        outExp = newExp;
        outMant = mant >>> 13;

        // 舍入到最近偶数
        int roundBit = (mant >>> 12) & 1;
        int sticky = mant & 0xFFF;

        if (roundBit == 1 && (sticky != 0 || (outMant & 1) == 1)) {
          outMant++;
          if (outMant > 0x3FF) {
            // 尾数溢出，增加指数
            outMant = 0;
            outExp++;
            if (outExp >= 31) {
              outExp = 0x1F; // 溢出到 Inf
            }
          }
        }
      }
    }

    return (short) (outSign | (outExp << 10) | outMant);
  }

  /**
   * 将 float16 位模式转换为 float32.
   *
   * <p>此转换对所有 float16 值都是精确的。
   */
  private static float float16BitsToFloat(short bits16) {
    int bits = bits16 & 0xFFFF;
    int sign = (bits >>> 15) & 0x1;
    int exp = (bits >>> 10) & 0x1F;
    int mant = bits & 0x3FF;

    int outBits = sign << 31;

    // NaN 或 Inf
    if (exp == 0x1F) {
      outBits |= 0xFF << 23;
      if (mant != 0) {
        // NaN：提升尾数
        outBits |= mant << 13;
      }
    }
    // 零或次正规
    else if (exp == 0) {
      if (mant == 0) {
        // 有符号零
        // outBits 已经正确
      } else {
        // 次正规：归一化
        // 值 = (-1)^S * 2^-14 * (mant / 1024)
        // 需要找到隐式 1 的位置
        int shift = Integer.numberOfLeadingZeros(mant) - 22; // 32 - 10
        mant = (mant << shift) & 0x3FF;
        int newExp = 1 - 15 - shift + 127;
        outBits |= newExp << 23;
        outBits |= mant << 13;
      }
    }
    // 正规数
    else {
      outBits |= (exp - 15 + 127) << 23;
      outBits |= mant << 13;
    }

    return Float.intBitsToFloat(outBits);
  }

  // ========== 分类方法 ==========

  /** 判断是否为 NaN. */
  public boolean isNaN() {
    return (bits & EXP_MASK) == EXP_MASK && (bits & MANT_MASK) != 0;
  }

  /** 判断是否为无穷大（正或负）. */
  public boolean isInfinite() {
    return (bits & EXP_MASK) == EXP_MASK && (bits & MANT_MASK) == 0;
  }

  /** 判断是否为有限值（非 NaN 且非无穷大）. */
  public boolean isFinite() {
    return (bits & EXP_MASK) != EXP_MASK;
  }

  /** 判断是否为零（正零或负零）. */
  public boolean isZero() {
    return (bits & (EXP_MASK | MANT_MASK)) == 0;
  }

  /** 判断是否为正规数（非零、非次正规、非无穷大、非 NaN）. */
  public boolean isNormal() {
    int exp = bits & EXP_MASK;
    return exp != 0 && exp != EXP_MASK;
  }

  /** 判断是否为次正规数. */
  public boolean isSubnormal() {
    return (bits & EXP_MASK) == 0 && (bits & MANT_MASK) != 0;
  }

  /** 判断符号位（true 表示负数或负零）. */
  public boolean signbit() {
    return (bits & SIGN_MASK) != 0;
  }

  // ========== 算术运算（通过 float32 提升）==========

  /**
   * 加法运算.
   *
   * @param other 另一个 Float16 值
   * @return 相加结果
   */
  public Float16 add(Float16 other) {
    return valueOf(floatValue() + other.floatValue());
  }

  /**
   * 减法运算.
   *
   * @param other 另一个 Float16 值
   * @return 相减结果
   */
  public Float16 subtract(Float16 other) {
    return valueOf(floatValue() - other.floatValue());
  }

  /**
   * 乘法运算.
   *
   * @param other 另一个 Float16 值
   * @return 相乘结果
   */
  public Float16 multiply(Float16 other) {
    return valueOf(floatValue() * other.floatValue());
  }

  /**
   * 除法运算.
   *
   * @param other 另一个 Float16 值
   * @return 相除结果
   */
  public Float16 divide(Float16 other) {
    return valueOf(floatValue() / other.floatValue());
  }

  /**
   * 取反运算.
   *
   * @return 取反结果
   */
  public Float16 negate() {
    return fromBits((short) (bits ^ SIGN_MASK));
  }

  /**
   * 绝对值运算.
   *
   * @return 绝对值
   */
  public Float16 abs() {
    return fromBits((short) (bits & ~SIGN_MASK));
  }

  // ========== Number 实现 ==========

  @Override
  public float floatValue() {
    return float16BitsToFloat(bits);
  }

  @Override
  public double doubleValue() {
    return floatValue();
  }

  @Override
  public int intValue() {
    return (int) floatValue();
  }

  @Override
  public long longValue() {
    return (long) floatValue();
  }

  @Override
  public byte byteValue() {
    return (byte) floatValue();
  }

  @Override
  public short shortValue() {
    return (short) floatValue();
  }

  // ========== 比较方法 ==========

  /**
   * 值相等比较（IEEE 754 语义：NaN != NaN，+0 == -0）.
   *
   * @param other 另一个 Float16 值
   * @return 是否数值相等
   */
  public boolean equalsValue(Float16 other) {
    if (isNaN() || other.isNaN()) {
      return false;
    }
    if (isZero() && other.isZero()) {
      return true;
    }
    return bits == other.bits;
  }

  @Override
  public int compareTo(Float16 other) {
    return Float.compare(floatValue(), other.floatValue());
  }

  /**
   * 对象相等比较（位模式相等）.
   *
   * <p>注意：此方法使用位模式相等，与 {@link #equalsValue(Float16)} 不同。 对于 IEEE 754 数值相等，请使用 equalsValue()。
   */
  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof Float16)) {
      return false;
    }
    Float16 other = (Float16) obj;
    return bits == other.bits;
  }

  @Override
  public int hashCode() {
    return Short.hashCode(bits);
  }

  @Override
  public String toString() {
    return Float.toString(floatValue());
  }
}

File: dart/packages/fory/lib/src/datatype/float16.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'dart:math' as math;
import 'dart:typed_data';

import 'float32.dart' show Float32;
import 'fory_fixed_num.dart';
import 'int16.dart' show Int16;
import 'int32.dart' show Int32;
import 'int8.dart' show Int8;

/// Float16: 16-bit floating point (IEEE 754 half precision)
/// Wraps a 16-bit integer representing the IEEE 754 binary16 format.
final class Float16 extends FixedNum {
  /// The raw 16-bit integer storage.
  final int _bits;

  /// Masks the bits to ensure 16-bit range.
  static const int _mask = 0xFFFF;

  // --- Constants ---
  static const int _exponentBias = 15;
  static const int _maxExponent = 31; // 2^5 - 1
  
  // Bit placeholders
  static const int _signMask = 0x8000;
  static const int _exponentMask = 0x7C00;
  static const int _mantissaMask = 0x03FF;

  // --- Public Constants ---
  static const Float16 positiveZero = Float16.fromBits(0x0000);
  static const Float16 negativeZero = Float16.fromBits(0x8000);
  static const Float16 positiveInfinity = Float16.fromBits(0x7C00);
  static const Float16 negativeInfinity = Float16.fromBits(0xFC00);
  static const Float16 nan = Float16.fromBits(0x7E00);

  static const double minValue = 6.103515625e-05;
  static const double minSubnormal = 5.960464477539063e-08;
  static const double maxValue = 65504.0;
  static const double epsilon = 0.0009765625;

  /// Constructs a [Float16] from a number. 
  /// Delegates to [Float16.fromDouble].
  factory Float16(num value) => Float16.fromDouble(value.toDouble());

  /// Constructs a [Float16] directly from raw bits.
  const Float16.fromBits(int bits) : _bits = bits & _mask, super();

  /// Converts a [double] to [Float16] using IEEE 754 half-precision rules (round-to-nearest-even).
  factory Float16.fromDouble(double value) {
     if (value.isNaN) {
       return Float16.nan;
     }

     final int doubleBits = _doubleToBits(value);
     final int sign = (doubleBits >> 63) & 0x1;
     final int rawExp = (doubleBits >> 52) & 0x7FF;
     final int rawMantissa = doubleBits & 0xFFFFFFFFFFFFF;

     // 1. Convert double exp to float16 exp
     // Double bias: 1023, Float16 bias: 15.
     // Shift: 1023 - 15 = 1008.
     int exp = rawExp - 1023 + _exponentBias;
     
     if (exp >= _maxExponent) {
       // Overflow or Infinity
        return rawExp == 2047 && rawMantissa != 0
            ? Float16.nan // Should have been caught by value.isNaN check above usually
            : (sign == 0 ? Float16.positiveInfinity : Float16.negativeInfinity);
     } else if (exp <= 0) {
       // Subnormal or Zero
       if (exp < -10) {
         // Too small for subnormal -> signed zero
         return sign == 0 ? Float16.positiveZero : Float16.negativeZero;
       }
       
       // Convert to subnormal
       return Float16.fromBits(_doubleToFloat16Bits(value));
     } else {
       // Normalized
       return Float16.fromBits(_doubleToFloat16Bits(value));
     }
  }

  static int _doubleToBits(double value) {
    var bdata = ByteData(8);
    bdata.setFloat64(0, value, Endian.little);
    return bdata.getUint64(0, Endian.little);
  }
  
  static double _bitsToDouble(int bits) {
      // Logic for converting float16 bits to double
      int s = (bits >> 15) & 0x0001;
      int e = (bits >> 10) & 0x001f;
      int m = bits & 0x03ff;
      
      if (e == 0) {
          if (m == 0) {
              // signed zero
              return s == 1 ? -0.0 : 0.0;
          } else {
              // subnormal
              return (s == 1 ? -1 : 1) * math.pow(2, -14) * (m / 1024.0);
          }
      } else if (e == 31) {
          if (m == 0) {
              return s == 1 ? double.negativeInfinity : double.infinity;
          } else {
              return double.nan;
          }
      } else {
          // normalized
          return (s == 1 ? -1 : 1) * math.pow(2, e - 15) * (1 + m / 1024.0);
      }
  }
  
  /// Helper to convert double to float16 bits with proper rounding
  static int _doubleToFloat16Bits(double val) {
      if (val.isNaN) return 0x7E00; // Canonical NaN
      
      // Check for zero
      if (val == 0.0) {
          return (1 / val) == double.negativeInfinity ? 0x8000 : 0x0000; 
      }
      
      int fbits = _doubleToBits(val);
      int sign = (fbits >> 63) & 1;
      int exp = (fbits >> 52) & 0x7FF; // 11 bits
      
      // Bias adjustment
      int newExp = exp - 1023 + 15;
      
      // Inf / NaN handled?
      if (exp == 2047) {
          // Infinity only (NaN handled at start)
          return (sign << 15) | 0x7C00;
      }
      
      if (newExp >= 31) {
           // Overflow to Infinity
           return (sign << 15) | 0x7C00;
      }
      
      if (newExp <= 0) {
          // Possible subnormal or zero
          if (newExp < -10) {
             // Underflow to zero
             return sign << 15;
          }
          
          double absVal = val.abs();
          if (absVal < minValue) {
              // It is subnormal
              // val = 0.m * 2^-14
              // m = val / 2^-14
              double m = absVal / math.pow(2, -14);
              // m is now in [0, 1). 
              // We want 10 bits of it.
              // bits = round(m * 1024)
              int mBits = (m * 1024).round();
              return (sign << 15) | mBits;
          }
      }
      
      // Normalized
      // We need to round the mantissa.
      
      int fullMantissa = (fbits & 0xFFFFFFFFFFFFF);
      // We want to reduce 52 bits to 10 bits.
      // So we shift right by 42.
      // The bits we shift out are the last 42 bits.
      // The 42nd bit (from right, 0-indexed) is the round bit.
      // The bits 0-41 are safety/sticky bits.
      
      int m10 = fullMantissa >> 42;
      int guard = (fullMantissa >> 41) & 1;
      int sticky = (fullMantissa & 0x1FFFFFFFFFF) != 0 ? 1 : 0;
      
      if (guard == 1) {
          if (sticky == 1 || (m10 & 1) == 1) {
              m10++;
          }
      }
      
      if (m10 >= 1024) {
          // Mantissa overflowed, increment exponent
          m10 = 0;
          newExp++;
          if (newExp >= 31) {
              return (sign << 15) | 0x7C00; // Inf
          }
      }
      
      return (sign << 15) | (newExp << 10) | m10;
  }

  /// Returns the raw 16-bit integer.
  int toBits() => _bits;

  /// Returns the value as a [double].
  double toDouble() => _bitsToDouble(_bits);

  /// Returns the underlying values as a [double].
  @override
  double get value => toDouble();

  // --- Classification ---
  bool get isNaN => (_bits & _exponentMask) == _exponentMask && (_bits & _mantissaMask) != 0;
  
  bool get isInfinite => (_bits & _exponentMask) == _exponentMask && (_bits & _mantissaMask) == 0;
  
  bool get isFinite => (_bits & _exponentMask) != _exponentMask;
  
  bool get isNormal => 
      (_bits & _exponentMask) != 0 && (_bits & _exponentMask) != _exponentMask;
      
  bool get isSubnormal => 
      (_bits & _exponentMask) == 0 && (_bits & _mantissaMask) != 0;
      
  bool get isZero => (_bits & 0x7FFF) == 0;
  
  bool get isNegative => (_bits & _signMask) != 0;
  
  int get sign => isNaN ? 0 : (isNegative ? -1 : 1); 

  // --- Arithmetic (Explicit) ---

  Float16 add(Float16 other) => Float16.fromDouble(toDouble() + other.toDouble());
  Float16 sub(Float16 other) => Float16.fromDouble(toDouble() - other.toDouble());
  Float16 mul(Float16 other) => Float16.fromDouble(toDouble() * other.toDouble());
  Float16 div(Float16 other) => Float16.fromDouble(toDouble() / other.toDouble());
  
  Float16 neg() => Float16.fromBits(_bits ^ _signMask);
  
  Float16 abs() => Float16.fromBits(_bits & 0x7FFF);

  // --- Comparisons ---
  
  /// Bitwise equality. +0 != -0, NaN == NaN (if payload matches).
  /// This is effectively `_bits == other._bits`.
  bool equalsValue(Float16 other) => _bits == other._bits;
  
  /// IEEE comparison.
  /// NaN != NaN. +0 == -0.
  bool ieeeEquals(Float16 other) => toDouble() == other.toDouble();
  
  bool lessThan(Float16 other) => toDouble() < other.toDouble();
  bool lessThanOrEqual(Float16 other) => toDouble() <= other.toDouble();
  bool greaterThan(Float16 other) => toDouble() > other.toDouble();
  bool greaterThanOrEqual(Float16 other) => toDouble() >= other.toDouble();
  
  static int compare(Float16 a, Float16 b) => a.toDouble().compareTo(b.toDouble());

  // --- Operators (Delegation) ---
  
  Float16 operator +(dynamic other) =>
      add(other is Float16 ? other : Float16(other));

  Float16 operator -(dynamic other) =>
      sub(other is Float16 ? other : Float16(other));

  Float16 operator *(dynamic other) =>
      mul(other is Float16 ? other : Float16(other));

  double operator /(dynamic other) =>
      toDouble() / (other is Float16 ? other.toDouble() : other);
      
  Float16 operator ~/(dynamic other) =>
      Float16((toDouble() ~/ (other is Float16 ? other.toDouble() : other)));

  Float16 operator %(dynamic other) =>
      Float16(toDouble() % (other is Float16 ? other.toDouble() : other));

  Float16 operator -() => neg();

  // Comparison Operators
  bool operator <(dynamic other) => toDouble() < (other is Float16 ? other.toDouble() : other);

  bool operator <=(dynamic other) => toDouble() <= (other is Float16 ? other.toDouble() : other);

  bool operator >(dynamic other) => toDouble() > (other is Float16 ? other.toDouble() : other);

  bool operator >=(dynamic other) => toDouble() >= (other is Float16 ? other.toDouble() : other);

  @override
  bool operator ==(Object other) {
    if (other is Float16) return _bits == other._bits; // Policy A: Bitwise equality
    return false;
  }

  @override
  int get hashCode => _bits.hashCode;

  // --- Type Conversions ---
  
  int toInt() => toDouble().toInt();
  
  Int8 toInt8() => Int8(toInt());
  Int16 toInt16() => Int16(toInt());
  Int32 toInt32() => Int32(toInt());
  
  Float32 toFloat32() => Float32(toDouble());
  
  // --- String Formatting ---
  
  String toStringAsFixed(int fractionDigits) => toDouble().toStringAsFixed(fractionDigits);
  String toStringAsExponential([int? fractionDigits]) => toDouble().toStringAsExponential(fractionDigits);
  String toStringAsPrecision(int precision) => toDouble().toStringAsPrecision(precision);
  
  @override
  String toString() => toDouble().toString();
}
